apps/hash-graph/lib/graph/src/snapshot.rs

```diff
@@ -176,57 +184,92 @@ where
+        chunk_size: usize,
     ) -> Result<(), SnapshotRestoreError> {
-        let mut snapshot = pin!(snapshot);
-        while let Some(entry) = snapshot.next().await {
-            let entry = entry.change_context(SnapshotRestoreError::Canceled)?;
-
-            match entry {
-                SnapshotEntry::Snapshot(global) => {
-                    ensure!(
-                        global.block_protocol_module_versions.graph
-                            == semver::Version::new(0, 3, 0),
-                        SnapshotRestoreError::Unsupported
-                    );
-                }
-                SnapshotEntry::DataType(data_type) => {
-                    tracing::trace!(
-                        "Inserting data type: {:?}",
-                        data_type.metadata.record_id.base_url
-                    );
-                }
-                SnapshotEntry::PropertyType(property_type) => {
-                    tracing::trace!(
-                        "Inserting property type: {:?}",
-                        property_type.metadata.record_id.base_url
-                    );
-                }
-                SnapshotEntry::EntityType(entity_type) => {
-                    tracing::trace!(
-                        "Inserting entity type: {:?}",
-                        entity_type.metadata.record_id.base_url
-                    );
-                }
-                SnapshotEntry::Entity(entity) => {
-                    tracing::trace!(
-                        "Inserting entity: {:?}",
-                        entity.metadata.record_id.entity_id.entity_uuid
-                    );
-                }
-            }
-        }
+        tracing::info!("snapshot restore started");
+
+        let (snapshot_record_tx, snapshot_record_rx) = restore::channel(chunk_size);
+
+        let read_thread = tokio::spawn(
+            snapshot
+                .map_err(|report| report.change_context(SnapshotRestoreError::Read))
+                .forward(
+                    snapshot_record_tx
+                        .sink_map_err(|report| report.change_context(SnapshotRestoreError::Buffer)),
+                ),
+        );
 
-        todo!("https://app.asana.com/0/0/1204216809501006/f")
+        let client = self
+            .0
+            .transaction()
+            .await
+            .change_context(SnapshotRestoreError::Write)?;
+
+        SnapshotRecordBatch::begin(&client)
+            .await
+            .change_context(SnapshotRestoreError::Write)?;
+
+        let client = snapshot_record_rx
+            .fold(
+                Ok(client),
+                |client, records: SnapshotRecordBatch| async move {
+                    let client = client?;
+                    records
+                        .write(&client)
+                        .await
+                        .change_context(SnapshotRestoreError::Write)?;
+                    Ok::<_, Report<SnapshotRestoreError>>(client)
+                },
```

Comment:
Actually, I'm not 100% about the behavior. I haven't used `try_fold` as it requires the `TryStream` instead of `Stream` (`Item = Result<...>`). I agree, that `try_fold` is easier to read. By mapping the stream I was able to use `try_fold` instead: [Use try_fold instead of fold](https://github.com/hashintel/hash/pull/2398/commits/9695a312394e94775a8b962a252cdeca1304f462)

Reply:
Thinking about it again you were probably right, that it would iterate the whole stream before actually emitting the first error.